### ISSUE TITLE
[Idea] VMAF ranges crf-search

### ISSUE THREAD

**Issue by Werve** [alexheretic/ab-av1#49]
I would like to share an idea for the crf-search part that I have encountered in prolonged use:

Sometimes the specified VMAF value is not found by the search, which then ends with "Error: Failed to find a suitable crf"
but the user may be fine with a nearby value.
For example. 
- I specify the VMAF value 98 
- but the best result found is 97 
-I am fine with that but that 97 found has an excessive estimated file size and the search does not continue to find a different crf with VMAF 97 because it was started for value 98

Basically a VMAF range would have to be accepted to perform the search (in this case it would be 98-97).

For quite some time I have been doing this simply with a bash script and crf-search but there is a lot of overhead as it has to repeat the search several times for different VMAF values even for crfs that have already been analyzed previously.

**Comment by alexheretic** [comment]
iiuc you want vmaf 98 but would be happy with 97, but only if 98 is impossible to satisfy.

To me the vmaf range arg idea doesn't really convey this. Or at least I don't find the --min-vmaf range an immediately understandable concept.

I'm not 100% sure how to convey this idea clearly in args. Maybe if we had an additional vmaf arg that is your "preferred" quality like `--preferred-vmaf=98 --min-vmaf=97`. 

Or alternatively `--min-vmaf=98 --failover-min-vmaf=97` meaning try again with this other min-vmaf.

I'm not sure myself.

**Comment by Werve** [comment]
> Or alternatively --min-vmaf=98 --failover-min-vmaf=97 meaning try again with this other min-vmaf.

Yes this is similar to the approach I used with the batch script redo the search by changing the min-VMAF value for all values within the range (e.g. range 98-96, means search for 98 then 97 then 96). 
It is rather inefficient though since it will also recalculate attempts with the same crf.
If instead you include logic in the program you could store the results of partial crf -> VMAF searches temporarily until the first completion of a search with min-VMAF present in the range.

I would solve with this pseudo algorithm.
For example, crf-search for the range 98-96:
- start the current crf-search algorithm with min-VMAF 98
- at each crf -> VMAF result (which are also written in CLI output) store them in an array
- if there is no result in the search with that VMAF (Error: Failed to find a suitable crf) start the crf-search algorithm again with the next VMAF in the range (in this example if you step up by one, 97)
- but in this new search exclude already tested crf -> saved VMAF encoding attempts that still do not match this new VMAF (97) so that no time is wasted in reencoding them having already the necessary values to be able to continue the binary search.

**Comment by alexheretic** [comment]
Yeah if it was built into ab-av1 it would re-use previously calculated crf results. So `--min-vmaf=98 --failover-min-vmaf=97` would try again, but re-using the work. Otherwise the feature wouldn't add much.

Another idea is to be able to save crf-search results in a file and re-use them on subsequent runs. It would only work using the same encoder & any encoder args though, so not as flexible an idea as it may first appear.

I also wonder if either approach is worth the extra complexity.

**Comment by Werve** [comment]
> I also wonder if either approach is worth the extra complexity.

Regarding storing crf results via file since it concerns that single video file (content) I think it is not worth the added code and checks.
A simple multiple invocation of crf-search with reading the values from an array if they are present before coding normally I think is the approach with least modification that may be worthy compared to the function performed.

Realistically I can assume that with a 98-96 VMAF range I can rule out 2 to 6 re-encodings in case the solution is found with 96 VMAF which on video files greater than 1h should roughly save at least twenty minutes (depends of course on the hardware in use, the settings chosen and particularly the complexity of the video content).

Regarding storing crf results via file since it concerns that single file (content) I think it is not worth the added code and steps.
A simple multiple invocation of crf-search with reading the values from an array if they are present before coding normally I think is the approach with least modification that may be worthy compared to the function performed.

Realistically I can assume that with a 98-96 VMAF range I can rule out 2 to 6 re-encodings in case the solution is found with 96 VMAF which on video files greater than 1h should roughly save at least twenty minutes (depends of course on the hardware in use, the settings chosen and particularly the complexity of the video content).

Personally I try to specify a VMAF value at which I do not visually notice a loss of quality to re-encode videos (>=96) and save space but preferring quality over weight for this the batch script I currently use starts the search with the highest VMAF value and then decreases.

**Comment by alexheretic** [comment]
Thanks for the further explanation, I'll have a think about how the cli args could work to provide this.

**Comment by Werve** [comment]
No problem, I share ideas happily, thank you for being available and open-minded 👍

**Comment by alexheretic** [comment]
I think I'm leaning toward the #59 style saving crf-search history. With that you can re-run with the same encoder opts and make use of previously calculated crf values while targeting whatever min-vmaf.

**Comment by Werve** [comment]
Certainly it is a useful addition to speed up the repeated search and would almost completely solve the question.
In order to implement the search in a VMAF range externally, as mentioned, for example with a powershell/batch script it would also be necessary to be able to read as a return value whether the search ended successfully or keep repeating the invocation with different --min-vmaf.
I had tried %errorlevel% but that doesn't seem to work.
Maybe is related to https://github.com/alexheretic/ab-av1/issues/58

**Comment by alexheretic** [comment]
> to be able to read as a return value whether the search ended successfully

The command should return non-zero (error) exit status if it fails to find a suitable crf or errors for any other reason.

**Comment by alexheretic** [comment]
With #92 merged you can now do something like
```sh
ab-av1 auto-encode -i v.mkv --min-vmaf 98 || ab-av1 auto-encode -i v.mkv --min-vmaf 97
```

**Comment by Werve** [comment]
I will use it that way next times, thanks for the update.
